pkg/virt-config/config-map.go

```diff
@@ -434,35 +415,88 @@ func setConfig(config *Config, configMap *k8sv1.ConfigMap) error {
+
+	overrides, err := json.Marshal(kvConfig)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(overrides, &config)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
 // getConfig returns the latest valid parsed config map result, or updates it
 // if a newer version is available.
 // XXX Rework this, to happen mostly in informer callbacks.
 // This will also allow us then to react to config changes and e.g. restart some controllers
-func (c *ClusterConfig) getConfig() (config *Config) {
+func (c *ClusterConfig) GetConfig() (config *v1.KubeVirtConfiguration) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
+	var configMap *k8sv1.ConfigMap
+	var kv *v1.KubeVirt
+	var resourceVersion string
+	useConfigMap := false
+
 	if obj, exists, err := c.configMapInformer.GetStore().GetByKey(c.namespace + "/" + ConfigMapName); err != nil {
-		log.DefaultLogger().Reason(err).Errorf("Error loading the cluster config from cache, falling back to last good resource version '%s'", c.lastValidConfig.ResourceVersion)
+		log.DefaultLogger().Reason(err).Errorf("Error loading the cluster config from cache, falling back to last good resource version '%s'", c.lastValidConfigResourceVersion)
 		return c.lastValidConfig
 	} else if !exists {
-		return c.defaultConfig
-	} else {
-		configMap := obj.(*k8sv1.ConfigMap)
-		if c.lastValidConfig.ResourceVersion == configMap.ResourceVersion ||
-			c.lastInvalidConfigResourceVersion == configMap.ResourceVersion {
+		kv = c.getConfigFromKubeVirtCR()
+		if kv == nil {
 			return c.lastValidConfig
 		}
-		config := defaultClusterConfig()
-		if err := setConfig(config, configMap); err != nil {
-			c.lastInvalidConfigResourceVersion = configMap.ResourceVersion
-			log.DefaultLogger().Reason(err).Errorf("Invalid cluster config with resource version '%s', falling back to last good resource version '%s'", configMap.ResourceVersion, c.lastValidConfig.ResourceVersion)
-			return c.lastValidConfig
-		}
-		log.DefaultLogger().Infof("Updating cluster config to resource version '%s'", configMap.ResourceVersion)
-		c.lastValidConfig = config
+
+		resourceVersion = kv.ResourceVersion
+	} else {
+		useConfigMap = true
+		configMap = obj.(*k8sv1.ConfigMap)
+		resourceVersion = configMap.ResourceVersion
+	}
+
+	// if there is a configuration config map present we should use its configuration
+	// and ignore configuration in kubevirt
+	if c.lastValidConfigResourceVersion == resourceVersion ||
+		c.lastInvalidConfigResourceVersion == resourceVersion {
+		return c.lastValidConfig
+	}
+
+	config = defaultClusterConfig()
+	var err error
+	if useConfigMap {
+		err = setConfigFromConfigMap(config, configMap)
+	} else {
+		err = setConfigFromKubeVirt(config, kv)
+	}
+
+	if err != nil {
+		c.lastInvalidConfigResourceVersion = resourceVersion
+		log.DefaultLogger().Reason(err).Errorf("Invalid cluster config with resource version '%s', falling back to last good resource version '%s'", resourceVersion, c.lastValidConfigResourceVersion)
```

Comment:
Can we differentiate in the error if it is an error coming from the CR or from the config-map? May make debugging easier.

Reply:
absolutely! i have added the resource type to this message as well as the "getting resource from cache" error messages.